runners/flink/runner/src/main/java/org/apache/beam/runners/flink/translation/FlinkStreamingTransformTranslators.java

```diff
@@ -284,8 +286,16 @@ public void flatMap(T s, Collector<WindowedValue<T>> collector) throws Exception
               }
             }).assignTimestampsAndWatermarks(new IngestionTimeExtractor<WindowedValue<T>>());
       } else {
-        source = context.getExecutionEnvironment()
-            .addSource(new UnboundedSourceWrapper<>(context.getPipelineOptions(), transform));
+        try {
+          UnboundedSourceWrapper<T, UnboundedSource.CheckpointMark> sourceWrapper =
+              new UnboundedSourceWrapper<>(
+                  context.getPipelineOptions(),
+                  (UnboundedSource) transform.getSource(),
+                  context.getExecutionEnvironment().getParallelism());
```

Comment:
This is correct, however, I'm performing the split in the constructor to ensure that all sources agree on how things are split and also to keep it consistent when it comes to checkpoints. Once Till has his work on live rebalancing in this might have to be replaced with the `maximum shards` parameter.